Update OCPP documentation (#64)
US-JOET/libocpp: doc/ocpp_201_README.md

![Github Actions](https://github.com/EVerest/libocpp/actions/workflows/build_and_test.yaml/badge.svg)

-----

# EVerest OCPP 2.0.1 Implementation (libocpp)
This is a C++ library implementation of OCPP for version 2.0.1
(see [OCPP protocols at OCA website](https://openchargealliance.org/protocols/open-charge-point-protocol/)).

It enables charging stations to communicate with cloud backends for remote control, monitoring and billing of charging processes.

Libocpp can be used for the communication of one charging station and multiple EVSE using a single websocket connection.

> OCPP 2.0.1 is currently under development.

Author: ```suggestion
# OCPP 2.0.1 in `libocpp`

The `ocpp::v201` namespace exposes a `ChargePoint` class capable of client-side OCPP 2.0.1 communications. It enables charging stations to communicate with a cloud backend over a websocket for the remote control of charging stations and administration of charging sessions.

> [!NOTE] For more information about the differences between versions of the Open Charge Point Protocol, please refer to the [Open Charge Alliance's website](https://openchargealliance.org/protocols/open-charge-point-protocol/).

The OCPP 2.0.1 implementation in `libocpp` is nearing completion, but there remain areas of active development. The OCA also does not currently plan to provide an official validation tool for all functionality within the protocol until the end of 2024. A high-level overview of supported functionality is provided [below](#supported-certification-profiles). For a more granular and up-to-date analysis of available OCPP 2.0.1 features in `libocpp`, please refer to the [OCPP 2.0.1 Status document](ocpp_201_status.md).
```

## Table of contents

- [Functionality Support](#functionality-support)
- [CSMS Compatibility](#csms-compatibility)
- [Integration with EVerest](#integration-with-everest)
- [Standalone Integration](#standalone-integration)
  - [Key Integration Points](#key-integration-points)
  - [Callbacks to Register](#callbacks-to-register)
  - [Event Handlers to Call](#event-handlers-to-call)
  - [Initialize the Database](#initialize-the-database)
  - [OCPP 2.0.1 Use Cases](#ocpp-201-use-cases)
- [Quickstart for OCPP 2.0.1](#quickstart-for-ocpp-201)
- [Build and Install `libocpp`](#build-and-install-libocpp)
- [Building with FetchContent Instead of EDM](#building-with-fetchcontent-instead-of-edm)
- [Unit Testing](#unit-testing)
- [Building the Doxygen Documentation](#building-the-doxygen-documentation)
- [Support for TPM Keys](#support-for-tpm-keys)
- [Support for `websocket++`](#support-for-websocket)

> - [Support for `iface`](#support-for-iface)

Author: ```suggestion
## Table of Contents

- [Supported Certification Profiles](#supported-certification-profiles)
- [CSMS Compatibility](#csms-compatibility)
- [Integration with EVerest](#integration-with-everest)
- [Standalone Integration](#standalone-integration)
  - [Key Integration Points](#key-integration-points)
  - [Callbacks to Register](#callbacks-to-register)
  - [Event Handlers to Call](#event-handlers-to-call)
  - [Initialize the Database](#initialize-the-database)
  - [OCPP 2.0.1 Use Cases in `libocpp`](#ocpp-201-use-cases-in-libocpp)
- [Quickstart for OCPP 2.0.1](#quickstart-for-ocpp-201)
- [Build and Install `libocpp`](#build-and-install-libocpp)
- [Building with FetchContent Instead of EDM](#building-with-fetchcontent-instead-of-edm)
- [Unit Testing](#unit-testing)
- [Building the Doxygen Documentation](#building-the-doxygen-documentation)
- [Support for TPM Keys](#support-for-tpm-keys)
- [Support for `websocket++`](#support-for-websocket)
- [Support for `iface`](#support-for-iface)
```

## Functionality Support

> The development of OCPP 2.0.1 implementation is in progress — [OCPP 2.0.1 status](/doc/ocpp_201_status.md). The following table shows the current support for OCPP 2.0.1 functionalities in libocpp:

Author: ```suggestion
## Supported Certification Profiles

Some optional OCPP 2.0.1 features are still being actively developed by the `libocpp` community. The following table shows which portions of the OCPP 2.0.1 implementation within `libocpp` are still being built out, broken down by [OCPP 2.0.1 Certification Profile](https://openchargealliance.org/certificationocpp/certification-ocpp-2-0-1/):
```

| OCPP 2.0.1 Functionality   | Supported                  |
| :------------------------- | :------------------------- |
| Core                       | :heavy_check_mark: Yes     |
| Advanced Security          | :construction: In Progress |
| Local Auth List Management | :hourglass: Planned        |
| Smart Charging             | :construction: In Progress |
| Advanced Device Management | :hourglass: Planned        |
| Advanced User Interface    | :hourglass: Planned        |
| Reservation                | :hourglass: Planned        |

> | ISO 15118 support          | :construction: In Progress |

Author: ```suggestion
| ISO 15118 Support          | :construction: In Progress |
```

| Whitepapers & Application Notes        | Supported                  |
| :------------------------------------- | :------------------------- |
| OCPP & California Pricing Requirements | :construction: In Progress |

## CSMS Compatibility

The ongoing development of OCPP 2.0.1 implementation is continuously tested against OCTT2 (OCPP Compliance Testing Tool 2), and has been tested against the following Charging Station Management Systems (CSMS):

| CSMS                                                                                 |
| :----------------------------------------------------------------------------------- |

> | Chargepoint                                                                          |

Author: ```suggestion
| [ChargePoint](https://www.chargepoint.com/businesses/software)                                                                          |
```

| [CitrineOS](https://lfenergy.org/projects/citrineos/)                                |

> | Current                                                                              |

Author: ```suggestion
| [Current](https://www.current.eco/)                                                                              |
```

| ihomer (Infuse CPMS)                                                                 |

> | Instituto Tecnológico de la Energía (ITE)                                            |

Author: ```suggestion
| [Instituto Tecnológico de la Energía (ITE) HySGrid+](https://www.ite.es/i-d-i/proyectos/hysgrid/)                                            |
```

| [MaEVe (Thoughtworks)](https://github.com/thoughtworks/maeve-csms)                   |
| [Monta](https://monta.com)                                                           |
| [Open Charging Cloud (GraphDefined)](https://github.com/OpenChargingCloud/WWCP_OCPP) |

> | Switch EV                                                                            |

Author: ```suggestion
| [Switch EV](https://www.switch-ev.com/platform)                                                                            |
```

> | SWTCH                                                                                |

Author: ```suggestion
| [SWTCH Energy](https://swtchenergy.com/)                                                                                |
```

## Integration with EVerest

The libocpp OCPP 2.0.1 module is automatically integrated within [everest-core](https://github.com/EVerest/everest-core) — the complete software stack for your charging station. 

> When you run libocpp with OCPP 2.0.1 through EVerest, the build process of everest-core handles the installation of all necessary dependencies. This includes initializing the device model database using the [config.json](config/v201/config.json) file.

Author: ```suggestion
## Integration with EVerest

EVerest provides a simplified OCPP 2.0.1 integration path through its `OCPP201` module.

When setting up a development environment (or building) EVerest through the [`everest-core`](https://github.com/EVerest/everest-core) project, a version of `libocpp` will be pulled down as a dependency. If you wish to use the `OCPP201` module, you will want to defined your charging station's Device Model in the [`config.json` file](config/v201/config.json) for the `libocpp` repository pulled in by `everest-core` as a dependency. This configuration will be incorporated into `everest-core` during the latter's build process.

Once `everest-core` has been built with your station-specific `libocpp` configuration, you can then specify and configure `OCPP201` as a module in the [runtime configuration](https://everest.github.io/nightly/general/04_detail_module_concept.html#wiring-it-all-together-the-run-configuration) of your EVerest deployment.
```

> ## Standalone Integration
> 
> If you wish to integrate libocpp's OCPP 2.0.1 implementation directly into your charging station software without using the full EVerest stack, you'll need to register several **callbacks** and implement **event handlers**. This allows libocpp to interact with your charging station according to OCPP requirements.

Author: ```suggestion
## Standalone Integration

OCPP is a protocol that affects, controls, and monitors many areas of a charging station's operation. The `libocpp` library is just the messenger for this protocol. It is intended to provide mechanisms for connecting to and authenticating with a CSMS, sending and receiving the OCPP messages that govern behaviors in the standard, and to the track state required for a charging station to conform to the protocol all while minimizing hardware- or implementation-specific functionality.

The actual substance of how a charging station reacts to an OCPP command (such as `ResetRequest`) or when it should consider sending a command to the CSMS (such as `NotifyChargingLimitRequest`) is left to the rest of the charging station's systems. This is done by providing means of (a) registering **callbacks** that can be triggered by a `libocpp` `ChargePoint` in response to certain events and (b) reacting to various **event handlers** defined on a `ChargePoint` in other areas of the charging station's codebase.

> [!IMPORTANT]
> Integrating this library with your charging station requires both (a) defining **callbacks** that enable control of your station by `libocpp` and (b) calling `libocpp` **event handlers** in your charging station's systems in response to new events and data in order to keep `libocpp` up to date on station information.
```

> ### Key Integration Points
> 
> 1. [**Callbacks**](#callbacks-to-register): Register these to allow libocpp to execute commands defined in OCPP (e.g., Reset.req or RemoteStartTransaction.req).
> 2. [**Event Handlers**](#event-handlers-to-call): Call these to notify libocpp of events, enabling the library to track the charging station's state and trigger appropriate OCPP messages (e.g., MeterValues.req, StatusNotification.req).
> 
> > [!note]
> > The public API for these integrations are found in the [ChargePoint](/include/ocpp/v201/charge_point.hpp).

Author: ```suggestion
### Terminology
Throughout this document and the `libocpp` codebase, the following conventions are followed:

- A **callback** is a function providing the actual station-specific implementation of an OCPP command. It allows a `libocpp` `ChargePoint` to control other systems within a charging station. By convention, each callback on a `ChargePoint` has a name suffixed with `_callback` (for instance, `unlock_connector_callback`). A `Callbacks` `struct` containing a complete set of desired callback implementations (as suitable `std::function` instances) should be provided to a 2.0.1 `ChargePoint` at construction time.

- An **event handler** is a public function defined on a `ChargePoint` that allows a charging station to update the state being tracked by the `ChargePoint` based on new information (meter values, charging session events, etc.) and (indirectly) send messages to a CSMS. By convention, the names of event handlers on the `ChargePoint` are each prefixed with `on_` (for instance, `on_meter_value`).

The complete set of callbacks and event handlers defined on an OCPP 2.0.1 `ChargePoint` can be viewed in the latter's [header file](/include/ocpp/v201/charge_point.hpp). Please use the [Doxygen documentation](#building-the-doxygen-documentation) as an additional source of information about the `ChargePoint` API.
```

| `transaction_event_response_callback`                   | Called when a transaction_event_response was received from the CSMS. |
| `unlock_connector_callback`                             | Used by libocpp to force unlock a connector |
| `update_firmware_request_callback`                      | Initiates a firmware update request. Triggers a security event notification if the certificate is Invalid or Revoked. |
| `validate_network_profile_callback`                     | Validates the submitted Network Profile. Is Rejected if:<br>• No callback registered to validate network profile<br>• CSMS attempted to set a network profile with a lower securityProfile<br>• CSMS attempted to set a network profile that could not be validated<br>• Network profile could not be written to the device model storage<br> |
| `variable_changed_callback`                             | Called when a variable has been changed by the CSMS |

### Event Handlers to Call
**table**

### Initialize the Database

Use [provided SQLite database](/doc/database_migrations.md) or implement your own storage drive.

> ### OCPP 2.0.1 Use Cases

Author: ```suggestion
## OCPP 2.0.1 Use Cases in `libocpp`

The OCPP 2.0.1 specification is broken down into concrete functional requirements using a three-tier system. At the top, there are sixteen **functional blocks** lettered A through P that describe the high-level desired capabilities of actors adhering to the standard. These functional blocks are then broken into **use cases** that describe different scenarios in which one or more charging stations, local controllers, and charging station management systems might find themselves. The required and optional behaviors of these systems in these scenarios are then declared as **functional requirements** within each use case.

This section includes an overview of how a subset of the OCPP 2.0.1 use cases that supported by `libocpp` are implemented in conjunction with callbacks provided by the charging station. We hope it serves as a useful guide to portions of the `libocpp` API and illustrates patterns of interaction between `libocpp` and other systems within a charging station.

> [!NOTE] Our goal is not to reproduce the entire OCPP 2.0.1 specification in this document. Rather we intend to illustrate how various use cases are realized using `ChargePoint` methods in `libocpp` in conjunction with callback implementations and event handlers.
>
> Please refer to the [OCPP 2.0.1 standard](https://openchargealliance.org/protocols/open-charge-point-protocol/#OCPP2.0.1) for more details. Functional blocks, use cases, and functional requirements are addressed in Part 2 of Edition 2.
```

> ### Event Handlers to Call
> Event handlers to call are found in the [ChargePoint](https://github.com/EVerest/libocpp/blob/2124d055f9cc8f253d3a2441ec2d70686174d0ed/include/ocpp/v201/charge_point.hpp#L116).

Author: ```suggestion
### Event Handlers

<!-- FIXME: Add a table of the current event handlers, or get rid of this section. -->
The complete set of event handlers defined on an OCPP 2.0.1 `ChargePoint` can be viewed in the latter's [header file](/include/ocpp/v201/charge_point.hpp). Please use the [Doxygen documentation](#building-the-doxygen-documentation) as an additional source of information about the `ChargePoint` API.
```

> Use provided [SQLite database](/doc/database_migrations.md) or implement your own storage drive.

Author: ```suggestion
To initialize the database where a `ChargePoint` will persist state, please use the provided [SQLite database](/doc/database_migrations.md) or implement your own storage drive.
```

| Clause | Use Case Name                     |
| :----- | :-------------------------------- |
| A      | Security                          |
| B      | Provisioning                      |
| C      | Authorization                     |
| D      | LocalAuthorizationList Management |
| E      | Transactions                      |
| F      | RemoteControl                     |
| G      | Availability                      |
| H      | Reservation                       |
| I      | TariffAndCost                     |
| J      | MeterValues                       |
| K      | SmartCharging                     |
| L      | FirmwareManagement                |
| M      | ISO 15118 CertificateManagement   |
| N      | Diagnostics                       |
| O      | DisplayMessage                    |

> | P      | DataTransfer                      |

Author: ```suggestion
| Functional Block | Functional Block Name               |
| :--------------- | :---------------------------------- |
| A                | Security                            |
| B                | Provisioning                        |
| C                | Authorization                       |
| D                | Local Authorization List Management |
| E                | Transactions                        |
| F                | Remote Control                      |
| G                | Availability                        |
| H                | Reservation                         |
| I                | Tariff and Cost                     |
| J                | Meter Values                        |
| K                | Smart Charging                      |
| L                | Firmware Management                 |
| M                | ISO 15118 Certificate Management    |
| N                | Diagnostics.                        |
| O                | Display Message                     |
| P                | Data Transfer                       |
```

> | Clause | Use Case Name                     |
> | :----- | :-------------------------------- |
> | A      | Security                          |
> | B      | Provisioning                      |
> | C      | Authorization                     |
> | D      | LocalAuthorizationList Management |
> | E      | Transactions                      |
> | F      | RemoteControl                     |
> | G      | Availability                      |
> | H      | Reservation                       |
> | I      | TariffAndCost                     |
> | J      | MeterValues                       |
> | K      | [SmartCharging](#smart-charging)  |
> | L      | FirmwareManagement                |
> | M      | ISO 15118 CertificateManagement   |
> | N      | Diagnostics                       |
> | O      | DisplayMessage                    |
> | P      | DataTransfer                      |

Author: ```suggestion
| Functional Block | Subject.                                    |
| :--------------- | :------------------------------------------ |
| A                | Security                                    |
| B                | Provisioning                                |
| C                | Authorization                               |
| D                | Local Authorization List Management         |
| E                | Transactions                                |
| F                | Remote Control                              |
| G                | Availability                                |
| H                | Reservation                                 |
| I                | Tariff and Cost                             |
| J                | Meter Values                                |
| K                | [Smart Charging](#smart-charging-use-cases) |
| L                | Firmware Management                         |
| M                | ISO 15118 Certificate Management            |
| N                | Diagnostics                                 |
| O                | Display Message                             |
| P                | Data Transfer                               |
```

Author: I'm not really sure this table is necessary anymore. It feels like a _non sequitur_ to me.

### Smart Charging

Smart Charging is a use case within OCPP.

> Work to fully support OCPP 2.0.1 Smart Charging is ongoing. Most functional requirements for General Smart Charging use cases (that is, K01–K10) are now supported. For an up-to-date overview of which features are currently supported, please refer to the [OCPP 2.0.1 Status](/doc/ocpp_201_status.md) document.

Author: ```suggestion
### Smart Charging Use Cases

The use cases within the Smart Charging functional block are subdivided into the following three categories of use cases:

1. General Smart Charging (Use Cases K01–K10)
2. External Charging Limit-based Smart Charging (K11–K14)
3. ISO 15118-based Smart Charging (K15–K17)

Support for General and External Charging Limit-based Smart Charging is largely complete, with ISO 15118-based Smart Charging under active development. For an up-to-date overview of exactly which features are currently supported as well as design decisions that have been made to address optional or ambiguous functional requirements, please refer to the [OCPP 2.0.1 Status document](ocpp_201_status.md).
```

| `TxProfileEvseHasNoActiveTransaction`                         | Happens when a `SetChargingProfileRequest` with a `TxProfile` is submitted and there is no transaction active on the specified EVSE. [K01.FR.09] |
| `TxProfileEvseIdNotGreaterThanZero`                           | `TxProfile` needs to have an `evseId` greater than 0. [K01.FR.16] |
| `TxProfileMissingTransactionId`                               | A `transactionId` is required for`SetChargingProfileRequest`s with a `TxProfile` in order to match the profile to a specific transation. [K01.FR.03] |
| `TxProfileTransactionNotOnEvse`                               | Happens when the provided `transactionId` is not known. [K01.FR.33] |
| `TxProfileConflictingStackLevel`                              | Happens when a TxProfile has a stackLevel and transactionId combination already exists in a TxProfile with a different id in order to ensure that no two charging profiles with same stack level and purpose can be valid at the same time. [K01.FR.39] |

> -----

Author: ```suggestion
```

        end
    end

    ChargePoint->>+EvseManager : does_evse_exist(call.msg.evseId)
    EvseManager-->>-ChargePoint : bool
    rect Red 
        break EVSE does not exist
            ChargePoint-->>CSMS : EvseDoesNotExist CallError
        end
    end

    ChargePoint->>+SmartChargingHandler : get_valid_profiles(call.msg.evseId)

    SmartChargingHandler-->>-ChargePoint : vector<ChargingProfile>

    ChargePoint->>+SmartChargingHandler : calculate_composite_schedule<br/>(vector<ChargingProfile, now, msg.duration, evseId, call.msg.chargingRateUnit)

    loop ExternalConstraints, Max, TxDefault, and Tx Profiles
        SmartChargingHandler->>+Profile: calculate_composite_schedule(profiles)
        Profile-->>-SmartChargingHandler: composite_schedule 
    end

    note right of SmartChargingHandler: Create consolidated CompositeSchedule<br />from all 4 Profile types

    SmartChargingHandler->>+Profile: calculate_composite_schedule(ExternalConstraints, Max, TxDefault, Tx)
    Profile-->>-SmartChargingHandler: CompositeSchedule

    SmartChargingHandler-->>-ChargePoint: CompositeSchedule

    ChargePoint-->>-CSMS : GetCompositeScheduleResponse(CompositeSchedule)
```

> -----

Author: ```suggestion
```

#### K09 Get Charging Profiles

Returns to the CSMS the Charging Schedules/limits installed on a Charging Station based on the 
passed in criteria.

```mermaid
sequenceDiagram

    CSMS->>+ChargePoint: GetChargingProfiles(criteria)

    ChargePoint->>+SmartChargingHandler : get_reported_profiles(criteria)

    loop filter ChargingProfiles
        SmartChargingHandler->>SmartChargingHandler: filter on ChargingProfile criteria 
    end

    SmartChargingHandler-->>-ChargePoint : Vector<Profiles>

    ChargePoint-->>CSMS : GetChargingProfilesResponse(profiles)

    alt no Profiles
    rect Red
        ChargePoint-->>CSMS : GetChargingProfilesResponse(NoProfiles) 
        ChargePoint->>CSMS : return
    end
        
    else Profiles
        ChargePoint-->>CSMS : GetChargingProfilesResponse(Accepted)
    end

    ChargePoint->>ChargePoint : determine profiles_to_report

    ChargePoint-->>-CSMS : ReportChargingProfilesRequest(profiles_to_report)
```

> -----

Author: ```suggestion
```

#### K10 Clear Charging Profile

Clears Charging Profiles installed on a Charging Station based on the
passed in criteria.

```mermaid
sequenceDiagram

    CSMS->>+ChargePoint: ClearChargingProfileRequest(criteria)

    alt no Profiles matching criteria
    rect Red
        ChargePoint-->>CSMS : ClearChargingProfileResponse(Unknown) 

    end
        
    else found matching Profiles
        ChargePoint-->>-CSMS : ClearChargingProfileResponse(Accepted)
    end    
```

> -----

Author: ```suggestion
```

## Quickstart for OCPP 2.0.1

> Libocpp provides a small standalone OCPP2.0.1 client that you can control using command line.

Author: ```suggestion
This section walks you through how to sanity check that your `libocpp` development environment is set up effectively. The process will involve the following:

1. Ensuring you've installed `libocpp`'s system dependencies
2. Standing up and configuring a CSMS (that is, an OCPP server) locally
3. Updating the default OCPP device model file
4. Updating the CSMS to make it aware of your test station's device model
5. Building `libocpp` as an OCPP 2.0.1 client executable
6. Running this freshly built binary

This approach can also be used later in development for testing an actual station's device model and configuration is valid in the eyes of a CSMS.
```

> Install the dependencies and libocpp as described in [Build and Install libocpp](#build-and-install-libocpp).

Author: ```suggestion
### Installing System Dependencies
Before proceeding further in this section, please ensure you have installed the system dependencies of `libocpp` as described in the [Build and Install `libocpp`](#build-and-install-libocpp) section.
```

Make sure you modify the following config entries in the [config.json](/config/v201/config.json) file according to the CSMS you want to connect to before executing make install.

```json
{
  "Internal": {
    "ChargePointId": "",
    "CentralSystemURI": ""
  }
}

> ```

Author: ````suggestion
### Configuring Your Charging Station

Before building the client binary, you will want to update the [`libocpp` configuration file](/config/v201/config.json) with accurate information about your desired charging station's hardware and how to access your CSMS. This configuration file includes an OCPP 2.0.1 device model for a charging station with two EVSEs, support for multiple network profiles, certificate-based authentication, etc. It also offers a host of `libocpp`-specific settings in an `InternalCtrlr` settings block.

Tailoring a device model to your station hardware and software is a large subject that deserves its own documentation. For now, we recommend at least validating the following about variables in the `InternalCtrlr`:

1. the `ChargePointId` variable matches the ID you've set (or plan to set) for this charging station in the CSMS, and
2. within the value of the `NetworkConnectionProfiles` variable, the `ocppCsmsUrl` field for the network connection profile you plan to use matches the websocket URL for your CSMS.

For completeness, we've also provided a table below of each `InternalCtrlr` variable as well as available metadata about the variable coming from the [`InternalCtrlr`'s JSON Schema file](/config/v201/component_config/standardized/InternalCtrlr.json).

<details><summary><code>InternalCtrlr</code> Variables</summary>
  <p>
    <markdown-accessiblity-table data-catalyst="">
      <table role="table">
        <thead>
          <tr>
            <th>Variable</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <th>InternalCtrlrEnabled</th>
            <td>boolean</td>
            <td>true</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>ChargePointId</th>
            <td>string</td>
            <td></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>NetworkConnectionProfiles</th>
            <td>string</td>
            <td>
              <code>"[{\"configurationSlot\": 1, \"connectionData\": {\"messageTimeout\": 30, \"ocppCsmsUrl\": \"ws://localhost:9000\", \"ocppInterface\": \"Wired0\", \"ocppTransport\": \"JSON\", \"ocppVersion\": \"OCPP20\", \"securityProfile\": 1}}]"</code>
            </td>
            <td>
              List of NetworkConnectionProfiles that define the functional and
              technical parameters of a communication link. Must be a (JSON) string
              with the format of <code>SetNetworkProfileRequest</code>.
            </td>
          </tr>
          <tr>
            <th>ChargeBoxSerialNumber</th>
            <td>string</td>
            <td></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>ChargePointModel</th>
            <td>string</td>
            <td></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>ChargePointSerialNumber</th>
            <td>string</td>
            <td></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>ChargePointVendor</th>
            <td>string</td>
            <td></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>FirmwareVersion</th>
            <td>string</td>
            <td></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>ICCID</th>
            <td>string</td>
            <td></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>IMSI</th>
            <td>string</td>
            <td></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>MeterSerialNumber</th>
            <td>string</td>
            <td></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>MeterType</th>
            <td>string</td>
            <td></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>SupportedCiphers12</th>
            <td>string</td>
            <td>
              <code>"ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:AES128-GCM-SHA256:AES256-GCM-SHA384"</code>
            </td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>SupportedCiphers13</th>
            <td>string</td>
            <td>
              <code>"TLS_AES_256_GCM_SHA384:TLS_AES_128_GCM_SHA256"</code>
            </td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>AuthorizeConnectorZeroOnConnectorOne</th>
            <td>boolean</td>
            <td>true</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>LogMessages</th>
            <td>boolean</td>
            <td>true</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>LogMessagesFormat</th>
            <td>string</td>
            <td><code>"log,html,security"</code></td>
            <td>
              Supported log formats are <code>console</code>, <code>log</code>,
              <code>html</code>, <code>console_detailed</code>,
              <code>callback</code> and <code>security</code>.
              <code>"security"</code> logs security events into a seperate logfile
            </td>
          </tr>
          <tr>
            <th>LogRotation</th>
            <td>boolean</td>
            <td>false</td>
            <td>Enable log rotation</td>
          </tr>
          <tr>
            <th>LogRotationDateSuffix</th>
            <td>boolean</td>
            <td>false</td>
            <td>
              Use a datetime suffix in log rotation files instead of the traditional
              <code>.0</code>, <code>.1</code>
            </td>
          </tr>
          <tr>
            <th>LogRotationMaximumFileSize</th>
            <td>integer</td>
            <td>0</td>
            <td>
              Maximum file size in bytes for the log file after which it will be
              rotated. Setting this to 0 disables log rotation.
            </td>
          </tr>
          <tr>
            <th>LogRotationMaximumFileCount</th>
            <td>integer</td>
            <td>0</td>
            <td>
              Maximum amount of files before rotated logs will be deleted. Setting
              this to 0 disables log rotation.
            </td>
          </tr>
          <tr>
            <th>SupportedChargingProfilePurposeTypes</th>
            <td>string</td>
            <td><code>"ChargePointMaxProfile,TxDefaultProfile,TxProfile"</code></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>MaxCompositeScheduleDuration</th>
            <td>integer</td>
            <td>31536000</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>NumberOfConnectors</th>
            <td>integer</td>
            <td>1</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>UseSslDefaultVerifyPaths</th>
            <td>boolean</td>
            <td>true</td>
            <td>Use default verify paths for validating CSMS server certificate</td>
          </tr>
          <tr>
            <th>VerifyCsmsCommonName</th>
            <td>boolean</td>
            <td>true</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>UseTPM</th>
            <td>boolean</td>
            <td>false</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>VerifyCsmsAllowWildcards</th>
            <td>boolean</td>
            <td>false</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>IFace</th>
            <td>string</td>
            <td><code>""</code></td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>EnableTLSKeylog</th>
            <td>boolean</td>
            <td>false</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>TLSKeylogFile</th>
            <td>string</td>
            <td>/tmp/ocpp_tlskey.log</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>OcspRequestInterval</th>
            <td>integer</td>
            <td>604800</td>
            <td>
              Interval in seconds used to request OCSP revocation status information
              on the CSO Sub-CA certificates
            </td>
          </tr>
          <tr>
            <th>WebsocketPingPayload</th>
            <td>string</td>
            <td>hello there</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>WebsocketPongTimeout</th>
            <td>integer</td>
            <td>5</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>MonitorsProcessingInterval</th>
            <td>integer</td>
            <td>1</td>
            <td>
              Defines the interval at which the periodic monitors will be processed,
              in seconds
            </td>
          </tr>
          <tr>
            <th>MaxCustomerInformationDataLength</th>
            <td>integer</td>
            <td>51200</td>
            <td>Maximum number of characters of Customer Information data</td>
          </tr>
          <tr>
            <th>V2GCertificateExpireCheckInitialDelaySeconds</th>
            <td>integer</td>
            <td>60</td>
            <td>
              Seconds to wait after boot notification to first check the V2G leaf
              certificate for expiration and potential renewal
            </td>
          </tr>
          <tr>
            <th>V2GCertificateExpireCheckIntervalSeconds</th>
            <td>integer</td>
            <td>43200</td>
            <td>
              Seconds between two checks for V2G leaf certificate expiration and
              potential renewal
            </td>
          </tr>
          <tr>
            <th>ClientCertificateExpireCheckInitialDelaySeconds</th>
            <td>integer</td>
            <td>60</td>
            <td>
              Seconds to wait after boot notification to first check the client
              certificate for expiration and potential renewal
            </td>
          </tr>
          <tr>
            <th>ClientCertificateExpireCheckIntervalSeconds</th>
            <td>integer</td>
            <td>43200</td>
            <td>
              Seconds between two checks for client certificate expiration and
              potential renewal
            </td>
          </tr>
          <tr>
            <th>UpdateCertificateSymlinks</th>
            <td>boolean</td>
            <td>false</td>
            <td>None Provided</td>
          </tr>
          <tr>
            <th>MessageQueueSizeThreshold</th>
            <td>integer</td>
            <td></td>
            <td>
              Threshold for the size of in-memory message queues used to buffer
              messages (and store e.g. while offline). If threshold is exceeded,
              messages  will be dropped according to OCPP specification to avoid
              memory issues.
            </td>
          </tr>
          <tr>
            <th>MaxMessageSize</th>
            <td>integer</td>
            <td>32000</td>
            <td>
              Maximum size in bytes for messages sent to the CSMS via websocket. If
              a message exceeds this size and is eligible to be split into multiple
              messages, it will be split. Otherwise, this value is ignored.
            </td>
          </tr>
          <tr>
            <th>SupportedCriteria</th>
            <td>string</td>
            <td><code>"Enabled,Active,Problem,Available"</code></td>
            <td>
              List of criteria supported for a get custom report.
              <code>Enabled,Active,Problem,Available</code>
            </td>
          </tr>
          <tr>
            <th>RoundClockAlignedTimestamps</th>
            <td>boolean</td>
            <td>false</td>
            <td>
              If enabled the <code>metervalues</code> configured with the
              <code>AlignedDataCtrlr</code> will be rounded to the exact time
              intervals
            </td>
          </tr>
          <tr>
            <th>ResumeTransactionsOnBoot</th>
            <td>boolean</td>
            <td>false</td>
            <td>
              If enabled the transactions that were active before shutdown will be
              resumed, if possible
            </td>
          </tr>
        </tbody>
      </table>
    </markdown-accessiblity-table>
  </p>
</details>
````
